app/configurator/components/filters.tsx

```diff
@@ -568,16 +570,49 @@ const TreeAccordion = ({
           {label}
         </Typography>
 
-        <SvgIcCheck
+        <Icon
+          name={
+            state === "SELECTED"
+              ? "check"
+              : state === "CHILDREN_SELECTED"
+              ? "indeterminate"
+              : "eye"
+          }
           className={classes.optionCheck}
-          style={{ visibility: selected ? "visible" : "hidden", marginTop: 8 }}
+          style={{
+            visibility: state === "NOT_SELECTED" ? "hidden" : "visible",
+            marginTop: 8,
+          }}
         />
       </TreeAccordionSummary>
       {children && <TreeAccordionDetails>{children}</TreeAccordionDetails>}
     </StyledAccordion>
   );
 };
 
+const areChildrenSelected = ({
+  children,
+  selectedValues,
+}: {
+  children: Maybe<HierarchyValue[]> | undefined;
+  selectedValues: HierarchyValue[];
+}): boolean => {
+  if (Array.isArray(children) && children.length > 0) {
+    const values = children.map((d) => d.value);
+    if (selectedValues.some((d) => values.includes(d.value))) {
+      return true;
+    } else {
+      return children
+        .map((d) =>
+          areChildrenSelected({ children: d.children, selectedValues })
+        )
+        .some((d) => d === true);
+    }
+  } else {
+    return false;
+  }
+};
```

Comment:
I think we could compute a tree of state, in one go we can compute all this tree, here we are doing multiple the same work (to compute the state of the parent, we must go over their children/grand-children, and then we process again the grand children to compute the state.

The easy way to have the same behavior would be to memoize.

```
const areChildrenSelectedMemoized = useMemo(() => {
	const areChildrenSelected = memoize((d) => {
	        const { children } = d
		if (Array.isArray(children) && children.length > 0) {
		  const values = children.map((d) => d.value);
		  if (selectedValues.some((d) => values.includes(d.value))) {
		    return true;
		  } else {
		    return children
		      .map((d) =>
		        areChildrenSelected(d)
		      )
		      .some((d) => d === true);
		  }
		} else {
		  return false;
		}
	})
       return do
}, [selectedValues])
```

Reply:
Yes, that's a good point, I also wasn't happy with re-computations – I'll introduce this change 🙇

Reply:
I guess here you'd need to have a second parameter to the memoize, `(d) => d.value`.